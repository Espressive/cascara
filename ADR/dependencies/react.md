# React

Our React version is tied to FE's version, we can bump only after they are doing bumping Ract. For now, all the following ecosystem members will be pinned to their current version:

- @types/react v16.8.0
- @types/react-dom v16.8.0
- react v16.13.1
- react-dom v16.13.1
- react-router v5.2.0
- react-router-dom v5.2.0
- react-scripts v4.0.3
- react-test-renderer v16.13.1
