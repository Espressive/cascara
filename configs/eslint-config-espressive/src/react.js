// "off" or 0 - turn the rule off
// "warn" or 1 - turn the rule on as a warning (doesn't affect exit code)
// "error" or 2 - turn the rule on as an error (exit code will be 1)

module.exports = {
  rules: {
    'react/boolean-prop-naming': 2,
    'react/default-props-match-prop-types': 2,
    'react/destructuring-assignment': 2,
    'react/jsx-fragments': 2,
    'react/jsx-no-duplicate-props': 2,
    'react/jsx-no-useless-fragment': 2,
    'react/jsx-pascal-case': 2,
    'react/jsx-sort-default-props': 2,
    'react/jsx-sort-props': 2,
    'react/no-children-prop': 2,
    'react/no-danger': 2,
    'react/no-find-dom-node': 2,
    'react/no-multi-comp': 2,
    'react/no-string-refs': 2,
    'react/no-this-in-sfc': 2,
    'react/no-typos': 2,
    'react/no-unused-prop-types': 2,
    'react/no-unused-state': 2,
    'react/prefer-stateless-function': 2,
    'react/self-closing-comp': 2,
    'react/sort-prop-types': 2,
    'react/style-prop-object': 2,
  },
};
