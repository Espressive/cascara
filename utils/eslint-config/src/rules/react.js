const { fixtures } = require('../utils/filePaths');

// For brevity, we start with the recomended defaults and extend with our own rules.
// Please make sure we are looking at any rules in the recommended set before adding
// them below. If we are overriding a recommended rule, we should make a comment for
// each rule.
//
// https://github.com/yannickcr/eslint-plugin-react

module.exports = {
  extends: ['plugin:react/recommended'],
  overrides: [
    {
      files: fixtures,
      rules: {
        'react/no-multi-comp': 'off',
        'react/prop-types': 'off',
      },
    },
  ],
  rules: {
    'react/boolean-prop-naming': 'error',
    'react/button-has-type': 'error',
    'react/default-props-match-prop-types': 'error',
    'react/destructuring-assignment': 'error',
    'react/forbid-foreign-prop-types': 'error',
    'react/forbid-prop-types': 'error',
    'react/function-component-definition': [
      'error',
      {
        namedComponents: 'arrow-function',
        unnamedComponents: 'arrow-function',
      },
    ],
    'react/jsx-boolean-value': 'error',
    'react/jsx-fragments': 'error',
    'react/jsx-key': 'error',
    'react/jsx-no-bind': 'error',
    'react/jsx-no-duplicate-props': 'error',
    'react/jsx-no-target-blank': 'error',
    'react/jsx-no-useless-fragment': 'error',
    'react/jsx-pascal-case': 'error',
    'react/jsx-sort-default-props': 'error',
    'react/jsx-sort-props': 'error',
    // 'react/jsx-wrap-multilines': ['error', {
    //   "arrow": "parens-new-line",
    //   "assignment": "parens-new-line",
    //   "condition": "ignore",
    //   "declaration": "parens-new-line",
    //   "logical": "ignore",
    //   "prop": "ignore",
    //   "return": "parens-new-line",
    // }],
    'react/no-access-state-in-setstate': 'error',
    'react/no-children-prop': 'error',
    'react/no-danger': 'error',
    'react/no-find-dom-node': 'error',
    'react/no-multi-comp': 'error',
    'react/no-render-return-value': 'error',
    'react/no-string-refs': 'error',
    'react/no-this-in-sfc': 'error',
    'react/no-typos': 'error',
    'react/no-unused-prop-types': 'error',
    'react/no-unused-state': 'error',
    'react/prefer-stateless-function': 'error',
    'react/require-render-return': 'error',
    'react/self-closing-comp': 'error',
    'react/sort-prop-types': 'error',
    'react/style-prop-object': 'error',
  },
};
