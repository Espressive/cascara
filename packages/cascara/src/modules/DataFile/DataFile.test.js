import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import cosmosFixtures, { displayProps, editingProps } from './DataFile.fixture';

// We cannot destructure during import because the default export in Cosmos
// multi-fixture files is an object so we need to import the fixtures first,
// then destructure them separately.
const { display, editing, displayNoLabel, editingNoLabel } = cosmosFixtures;

describe('DataFile', () => {
  // without ModuleSandbox will render the property information into a span
  describe('display', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(display).container;
    });

    test('snapshot', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <span> by default', () => {
      const input = screen.getByLabelText(displayProps.label);
      // Make sure the actual DOM element is not render an input
      expect(input.tagName).toMatch('SPAN');
      // Input file does not display an INPUT
      expect(input.classList.contains('Input')).toBe(false);
    });
  });

  describe('editing', () => {
    // We need a place to store the view for snapshot testing. This is not required when we are using `screen` directly from RTL.
    let view;

    beforeEach(() => {
      // Set the render container to our `view` so it is in scope for the snapshot test
      view = render(editing).container;
    });

    test('snapshot', () => {
      expect(view).toMatchSnapshot();
    });

    test('renders a <input date> by default', () => {
      const input = screen.getByLabelText(editingProps.label);
      // Check that we also use the correct type
      expect(input).toHaveAttribute('type', 'file');
    });

    test('selec a file', () => {
      const file = new File(['hello'], 'hello.png', { type: 'image/png' });
      const input = screen.getByLabelText(editingProps.label);
      userEvent.upload(input, file);
      expect(input.files[0]).toStrictEqual(file);
      expect(input.files.item(0)).toStrictEqual(file);
      expect(input.files).toHaveLength(1);
    });
  });

  describe('accessibility', () => {
    test('editing', () => {
      render(editing);

      const input = screen.getByLabelText(editingProps.label);
      // The label tag is the parent wrapper
      const label = input.closest('label');

      // Test is written this way to make sure we know that both values need to be the same.
      const linkedLabelValue = editingProps.label;

      // Verify label for attribute has linked value
      expect(label).toHaveAttribute(
        'for',
        expect.stringContaining(linkedLabelValue)
      );
      // Verify input id attribute has linked value
      expect(input).toHaveAttribute(
        'id',
        expect.stringContaining(linkedLabelValue)
      );
      // Check that the input does NOT have an aria-label defined because there is a label tag
      expect(input).not.toHaveAttribute('aria-label');
    });

    test('display no label', () => {
      // Make sure that the input is still accessible with label text even when we are not showing a label tag in tables
      render(displayNoLabel);
      const input = screen.getByLabelText(displayProps.label);
      expect(input).toBeDefined();
    });

    test('editing no label', () => {
      // Make sure that the input is still accessible with label text even when we are not showing a label tag in tables
      render(editingNoLabel);
      const input = screen.getByLabelText(editingProps.label);
      expect(input).toBeDefined();
    });
  });
});
