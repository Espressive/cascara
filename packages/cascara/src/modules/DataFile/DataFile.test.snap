// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataFile display snapshot 1`] = `
<div>
  <div
    class="File"
  >
    <span>
      <span
        class="LabelText"
      >
        Display
      </span>
      <span
        aria-label="Display"
        class="File"
      />
    </span>
  </div>
</div>
`;

exports[`DataFile editing snapshot 1`] = `
<div>
  <div
    class="File"
  >
    <label
      for="Editing"
    >
      <span
        class="LabelText"
      >
        Editing
      </span>
      <input
        class="Input"
        id="Editing"
        name="Editing"
        type="file"
      />
    </label>
  </div>
</div>
`;
